quinn-proto/src/endpoint.rs

```diff
@@ -748,17 +691,115 @@ impl fmt::Debug for Endpoint {
+    ///
+    /// Incoming stateless resets do not have correct CIDs, so we need this to identify the correct
+    /// recipient, if any.
+    connection_reset_tokens: ResetTokenTable,
+}
+
+impl ConnectionIndex {
+    /// Associate a connection with its initial destination CID
+    fn insert_initial(&mut self, dst_cid: ConnectionId, connection: ConnectionHandle) {
+        self.connection_ids_initial.insert(dst_cid, connection);
+    }
+
+    /// Associate a connection with its first locally-chosen destination CID if used, or otherwise
+    /// its current 4-tuple
+    fn insert_conn(
+        &mut self,
+        addresses: FourTuple,
+        dst_cid: ConnectionId,
+        connection: ConnectionHandle,
+    ) {
+        match dst_cid.len() {
+            0 => {
+                self.connection_remotes.insert(addresses, connection);
+            }
+            _ => {
+                self.connection_ids.insert(dst_cid, connection);
+            }
+        }
+    }
+
+    /// Add a new CID to an existing connection
+    fn insert_cid(&mut self, dst_cid: ConnectionId, connection: ConnectionHandle) {
+        self.connection_ids.insert(dst_cid, connection);
+    }
+
+    /// Discard a connection ID
+    fn retire(&mut self, dst_cid: &ConnectionId) {
+        self.connection_ids.remove(dst_cid);
+    }
+
+    /// Remove all references to a connection
+    fn remove(&mut self, conn: &ConnectionMeta) {
+        if conn.init_cid.len() > 0 {
+            self.connection_ids_initial.remove(&conn.init_cid);
+        }
+        for cid in conn.loc_cids.values() {
+            self.connection_ids.remove(cid);
+        }
+        self.connection_remotes.remove(&conn.addresses);
+        if let Some((remote, token)) = conn.reset_token {
+            self.connection_reset_tokens.remove(remote, token);
+        }
+    }
+
+    /// Find the existing connection that `datagram` should be routed to, if any
+    fn get(&self, addresses: &FourTuple, datagram: &PartialDecode) -> Option<ConnectionHandle> {
+        (datagram.dst_cid().len() != 0)
+            .then(|| self.connection_ids.get(datagram.dst_cid()))
+            .flatten()
+            .or_else(|| {
+                if datagram.is_initial() || datagram.is_0rtt() {
+                    self.connection_ids_initial.get(datagram.dst_cid())
+                } else {
+                    None
+                }
+            })
+            .or_else(|| {
+                if datagram.dst_cid().len() == 0 {
+                    self.connection_remotes.get(addresses)
+                } else {
+                    None
+                }
+            })
+            .or_else(|| {
+                let data = datagram.data();
+                if data.len() < RESET_TOKEN_SIZE {
+                    return None;
+                }
+                self.connection_reset_tokens
+                    .get(addresses.remote, &data[data.len() - RESET_TOKEN_SIZE..])
+            })
+            .cloned()
```

Comment:
In this unnested context, I think this could benefit from rewriting with early returns (and the `?` operator) and explicit matches rather than this combinator festival...

Reply:
Good call. We can't use `?` here because the desired fallback is always "try the next case", but it's still a fair bit more readable this way.